Colour handling improvements

Jermolene opened the pull request:
**UPDATE: this PR is deferred to v5.5.0** to allow more time to finalise the details. Background actions, filter tracking and media query tracking were cherry-picked into #9641 so they can ship in v5.4.0.

# Introduction

This PR brings several new features for end users:

* Automatically switching between a dark and light palette as the operating system setting changes (without making the wiki dirty)
* Customisation options for palettes. For example, users might choose a base hue, with the palette colours automatically adapting to it. Customisations are stored in a separate tiddler from the palette. This ensures that future updates to the palette itself will be honoured, and that the customisations can be reset.
* A generalisation of the dark vs. light mechanism to allow an arbitrary number of distinct schemes that are dynamically selected (e.g. a palette with separate schemes for night, morning, day and evening that change with the time of day)

There are also new capabilities for palette authors:

* Inheritance for palettes, making it easy to create chains of variants of a base palette
* Self-contained palettes that can contain both dark and light variants (or variants for any other custom scheme)

To make these possible, the PR also adds some general-purpose mechanisms:

* Several new filter operators for manipulating colour values, backed by the [color.js](https://colorjs.io/) library
* New `changecount` filter operator
* New `::defaultprefix` filter pragma for setting the default filter run prefix

(Background actions, the media query tracking mechanism and filter tracking were part of this work but have been split out into #9641 for v5.4.0.)

# Try the Demo

A preview build is available at https://deploy-preview-8702--tiddlywiki-previews.netlify.app/

# Changes to Palettes

Palette entries are now defined as filters that are evaluated, rather than wikitext that is wikified. This makes it possible to create palettes that reference and modify other colours, for example:

```
tiddler-controls-foreground-selected: [tf.interpolate-colours[background],[foreground],[0.9]]
```

Re-interpreting palette entries as filters would normally break `<<>>`. Instead, legacy entries are automatically converted to the updated form `[function[colour],[background]]`.

# Palette Compilation

The key idea underpinning these changes is a change to the way TiddlyWiki handles palettes. Previously, palette entries were named items containing a CSS colour string, a CSS colour token like `inherit`, or a `<<>>` reference to another entry - so entries had to be wikified before use. That is limiting and provides no viable path to the complex colour manipulations above.

The fix is compiled palettes: when switching to a new palette, its colours are "compiled" to raw CSS colour values (typically `#rrggbbaa`) in the system tiddler `$:/temp/palette-colours`, by an action procedure invoked at startup and on palette switch. This lets palette entries be used directly without wikification.

Because palette recompilation is driven by filter tracking, any code that changes `$:/palette` automatically triggers recompilation, so there should be no backwards-compatibility issues.

This change also lets the `<<colour>>` procedure become a function, so it can be used as the value of a style attribute.

# Automatic Palette Readability Tests

Palettes can include readability tests as special palette entries whose names start with `?`. The associated filter returns nothing if there are no errors, or a textual error message otherwise. Results are shown at the bottom of the palette chooser. For example:

```
?base-background-ink: [tf.check-colour-contrast[base-background],[base-ink],[45]]
```

Sample output:

```
base-paper-ink: 41.915: base-paper/base-ink contrast is too low
code-contrast: 28.565: code-background/code-foreground contrast is too low
```

# Colour Manipulation

We need a colour library that can calculate colour variants. Only [color.js](https://colorjs.io/) met the requirements of working with P3 colours and the OKLCH colour space. It also includes a CSS colour string parser that replaces the simple one TiddlyWiki has always used.

New filter operators: `colour-lighten`, `colour-darken`, `colour-set-alpha`, `colour-get-oklch`, `colour-set-oklch`, `colour-contrast`, `colour-best-contrast`, `colour-interpolate`.

# `changecount` Operator and `::defaultprefix` Pragma

* `changecount` - a new filter operator returning the change count for each title in the list.
* `::defaultprefix:<prefix>` - a new filter pragma that sets the default filter run prefix for subsequent runs in the filter. This is used internally so that the multi-value widgets (`genesis`, `setmultiplevariables`, `action-setmultiplefields`, `action-sendmessage`) preserve duplicate/positional alignment between their names and values.

Computed colour helpers are built from these operators using the `let` filter run prefix (the `=>variable` syntax) to hold intermediate values. For example, the interpolation helper used by the palettes:

```
\function tf.interpolate-colours(paletteEntryA,paletteEntryB,weight)
[function[colour],<paletteEntryA>] =>colourA [function[colour],<paletteEntryB>] =>colourB [<weight>colour-interpolate:oklch<colourA>,<colourB>]
\end tf.interpolate-colours
```

# Backwards Compatibility

* The previous content of `$:/PaletteManager` moves into `$:/PaletteEditor`, and `$:/PaletteManager` is repurposed as the control panel palette switcher
* `$:/config/DefaultColourMappings/` now only supports CSS colours, not indirections via `<<>>` or `[function[colour],[X]]`
* The internal CSS colour parser is now backed by color.js. The `$tw.utils.parseCSSColor` compatibility shim is retained and continues to return RGB components in the 0-255 range (alpha 0-1), so existing callers are unaffected.

# References

* Media query tracker and background actions were originally developed in #8555 and split out into #9641
* #5236 discusses dynamically computed colours, including a [demo](https://github.com/TiddlyWiki/TiddlyWiki5/issues/5236#issuecomment-890428532) from @cdruan
* Colour macro as an attribute value - https://talk.tiddlywiki.org/t/how-to-use-the-colour-macro-as-an-attribute-value/10051
* Improvements to colour handling / palette previews - https://talk.tiddlywiki.org/t/idea-approach-to-palettes-and-themes/9431

# Progress

* Documentation
  * [x] End-user: palette chooser UI / settings panel
  * [x] End-user: automatic dark/light switching
  * [x] End-user: palettes vs. themes and best practices
  * [x] Palettes and colour schemes (authoring)
  * [x] Palette inheritance
  * [x] Readability tests
  * [x] Update `ColourPalettes` concept tiddler for the compiled-palette model
  * [x] Update `ColourMacro` for the function change
  * [x] `colour-get-oklch`, `colour-set-oklch`, `colour-interpolate` operators
  * [x] `colour-lighten`, `colour-darken`, `colour-set-alpha`, `colour-contrast`, `colour-best-contrast` operators
  * [x] `changecount` operator
  * [x] `::defaultprefix` pragma
  * [x] Colour Spaces concept
  * [x] Adopt palettes from https://yatagarasu.tiddlyhost.com/
* Code
  * [x] Remove duplicate `$tw.perf` initialisation in `load-modules.js`
  * [x] Restore 0-255 range in the `$tw.utils.parseCSSColor` compatibility shim
  * [x] Review fixes: `filterTiddlers` call signature in the multi-value widgets; scheme-suffixed entry compilation and preview lookups; editor iframe colour scheme; guards for achromatic hues and unparseable colours; cycle protection in palette previews; recursive entry definitions reported via the readability test mechanism
  * [x] `colour-composite` operator, used to fix the contrast readability tests for translucent colours (with docs)
  * [x] Palette customisations stored in a per-palette layer tiddler (`$:/config/PaletteCustomisations/<palette>`) that imports the palette itself: non-destructive, resettable, and plugin palettes continue to receive updates. Reset buttons in the chooser and editor; overridden entries marked in the editor
  * [x] Vanilla fallback when `$:/palette` names a missing tiddler
  * [x] Final review pass: coverage audit (suite grown to 1,579 specs incl. JS specs for the colour utilities), docs consistency, browser test timeout raised for CI
  * [x] Convert the palettes-2026 plugin to the new system: shared hidden base palette, ten `base-*` parameters per palette, settings panel via field inheritance 
* Remaining decisions
  * [x] `:DeltaPhi` suffix on the contrast operators is currently ignored: implement or remove from callers
  * [x] `::defaultprefix` position sensitivity: resolved by restricting pragmas to the start of the filter, with generic Filter Pragmas docs and an impact note for titles beginning `::`
  * [x] Add impact note: the `<<colour x>>AA` hex-alpha-suffix idiom no longer works (use `colour-set-alpha`)
  * [x] Add change note for the `lingo` macro no longer setting `currentTiddler`
  * [x] `$:/config/DefaultColourMappings/` fallback is unused by core: document as third-party compatibility or remove
  * [x] Remove dead code: `actions-switch-colour-palette`; simplify the suffix fallback expression in `filter.js`/`subfilter.js`
  * [x] Merge master into the branch to clean the PR diff
  * [ ] Remove `$:/PaletteManager` from the tw5.com and prerelease `$:/DefaultTiddlers` before release (present deliberately for the preview build)

pmario commented on themes/tiddlywiki/vanilla/base.tid:
```diff
@@ -2853,19 +2867,22 @@ a.tc-tiddlylink.tc-plugin-info:hover > .tc-plugin-info-chunk .tc-plugin-info-sta
 	color: <<colour foreground>>;
 }
 
-.tc-chosen > .tc-tiddlylink:before {
```
IMO we need to keep these configuration for backwards compatibility. We can add a "deprecated" comment, but plugin authors may use these styles. 

I'll need to make some more tests, with the palette-manager edition.

pmario replied:
@Jermolene -- The tools-buttons: theme, layout, storyview and palette use this CSS. So IMO we can not remove it

pmario replied:
All the "chosen" indicators fail now.

![image](https://github.com/user-attachments/assets/ff46463d-babd-4067-92ec-0052594dc67d)

Jermolene replied:
Thanks @pmario, fixed in c6074402bbdcb8ef04c575090499874d02339a56

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Add expertimental colour-oklch operator

pmario commented on core/wiki/paletteswitcher.tid:
```diff
@@ -2,15 +2,18 @@ title: $:/snippets/paletteswitcher
 
```
Can you add indentation, so it's easier to read

Jermolene replied:
Good point will do

pmario commented on core/modules/utils/dom/color.js/tiddlywiki.files:
```diff
@@ -0,0 +1,21 @@
+			"fields": {
+				"type": "application/javascript",
+				"title": "$:/core/modules/utils/dom/color.js",
+				"module-type": "library"
```
IMO we should add 2 more fields: `version: 0.5.2` and `repo: https://github.com/color-js/color.js` to make it easier for us to maintain

Jermolene replied:
Hi @pmario if we were going to add those fields then we should do so consistently for all the 3rd party libraries in the core.

pmario replied:
I think that would be a good idea and we have to start somewhere.

kookma commented:
Is there a demo to test colour handling improvement?

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Indentation for palette switcher

Jermolene commented:
> Is there a demo to test colour handling improvement?

Hi @kookma I've posted a preview to [tiddlyhost](https://colour-improvements-pr8702.tiddlyhost.com/#%24%3A%2Fcore%2Fui%2FControlPanel%2FPalette)

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Restore default styling for chosen chooser item

See https://github.com/TiddlyWiki/TiddlyWiki5/pull/8702#discussion_r1816584692

Jermolene committed:
Remove logging

Jermolene committed:
Testing Netlify CI

Jermolene committed:
Testing CI

Jermolene committed:
Remove wikify operator and refactor palette preview

The implementation of the palette preview is much less elegant like this, but it does work

Jermolene committed:
Make preview templates tag driven and extensible

Jermolene committed:
Introduce new static palette architecture

Jermolene commented:
The latest update introduces the idea of compiling palettes into static palettes at the time the palette is switched. The details are covered in the OP. The backwards compatibility story is pretty good, and almost everything will continue to work as usual, except for custom colour palette switches. There are definitely still plenty of more minor backwards compatibility implications to think about, but I'm excited for how simple this makes everything.

kookma commented:
Thank you @Jermolene 



> The latest update introduces the idea of compiling palettes into static palettes at the time the palette is switched.

I opened the latest build, it seems they still use the colour macro. To make sure I switched several palettes and then bacl to Vanilla, then click on Show Editor. For example for Vanilla palette I see:

diff-delete-foreground: `<<colour foreground>>`

Jermolene commented:
> I opened the latest build; it seems they still use the color macro. To make sure I switched several palettes and then back to Vanilla, then click on Show Editor. For example, for Vanilla palette I see:

The existing palettes haven't changed, and are still compiled into static palettes by wikification. Have a look at the new TwentyTwenties palette. It is a new modern palette, which is compiled into a static palette by filter evaluation.

Jermolene committed:
Avoid redefining the colour function by adding a configuration variable

pmario commented:
@Jermolene ... Since you did already add a new field to the "modern" palette tiddlers, I would like to request one more.

The `origin` or `parent` field, which links to the "parent" palette if a palette is "cloned" with the palette editor. This allows developers to implement an **undo function** in the palette editor, since we do know, where to look up the "old" value.

pmario commented on core/ui/Palettes/Preview/SideBar/Search.tid:
```diff
@@ -0,0 +1,8 @@
+tags: $:/tags/Preview/SideBar
+
+\whitespace trim
+<$colour.div class="tc-palette-preview-thumbnail-sidebar-search" styleName="background-color" styleColour="background">
```
If parameters are named `style*` they immediately raise red flags in my mind. I associate those parameters with hardcoded STYLE tags in HTML. Hardcoded styles make colour customizing extremely hard. 

So I would like different widget parameter names if possible. 

I have to say I did not read the whole code yet. Only saw these parameters and go "triggered"

pmario replied:
>In any case, this custom widget is actually going to be removed soon

OK. After saving the post, I actually thought that they may be assigned to styles. -- If it is removed I'm relieved.

Jermolene replied:
Thanks @pmario

Jermolene committed:
Support for imported palettes

Also consolidate the palette entries into a temporary palette before compiling them

Note that imported palettes is not currently recursive

pmario commented on core/wiki/macros/CSS.tid:
```diff
@@ -1,39 +1,76 @@
-	<$list filter="[<inputTitle>indexes[]sort[]]" variable="colour-name">
-		<$let colour-filter={{{ [<inputTitle>getindex<colour-name>] }}}>
+\procedure actions-compile-palette-filtered(tempPalette,outputTitle)
+<!-- Make the colour function use the input palette -->
```
Is it true, that procedures are a little bit more performant it they are outside the definition?

Jermolene replied:
Perhaps marginally because there is the possibility of caching the parse tree. But in this case this is an action that is performed once, and not something that is repeated with each refresh cycle, and so performance is not the primary concern.

kookma commented:
Is there any plan to be able to use css variable like (https://developer.mozilla.org/en-US/docs/Web/CSS/Using_CSS_custom_properties)?

As this is an overhaul of palettes, I may recommend seeing the possibility to extend the color manipulation to CSS variables. There is some good CSS framework like bootstrap, Tailwind, Pure, ... which can be used in TiddlyWiki. So, I may recommend considering this possibility.

Jermolene committed:
Fix preview of filtered palettes

pmario commented:
The currently active palette is very hard to spot. So could you add this CSS, which would make it much more visible

```
.tc-chooser-cards .tc-chooser-item.tc-chosen {
  outline: 2px solid <<colour primary>>;
}
```

![image](https://github.com/user-attachments/assets/628b4689-563d-4535-b8c8-91905ee289b3)

Jermolene committed:
Improve chooser chosen item highlight

Jermolene committed:
Another fix for filtered palette previews

pmario commented on core/wiki/paletteswitcher.tid:
```diff
@@ -1,19 +1,26 @@
+		<$list filter="[all[shadows+tiddlers]tag[$:/tags/Palette]sort[name]]">
+			<$set name="cls" filter="[all[current]prefix{$:/palette}]" value="tc-chooser-item tc-chosen" emptyValue="tc-chooser-item">
+				<div class=<<cls>>>
+					<$link to={{!!title}}>
```
This suggestion will add an alias-label

```suggestion
	<$link to={{!!title}} aria-label=`${[[$:/language/ControlPanel/Palette/Caption]get[text]]}$ - ${[all[current]get[name]]}$. ${[all[current]get[description]]}$`>
```

Jermolene committed:
Add aria labels to palette switcher

Co-authored-by: Mario Pietsch <pmariojo@gmail.com>

Jermolene commented:
Thanks @pmario

pmario commented:
@Jermolene .. Does this PR already have an `contrastcolor` operator? That can be used in eg. tag-pills to calculate the text colour based on the background colour. I would like to test that one.

Jermolene commented:
> @Jermolene .. Does this PR already have an `contrastcolor` operator? That can be used in eg. tag-pills to calculate the text colour based on the background colour. I would like to test that one.

The existing contrastcolor operator works with the updated CSS parser, but I haven't otherwise made any improvements in that area.

My focus at the moment is still on getting the previews to work properly for filtered palettes, and finishing up the importing mechanism.

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Palette tweaks

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Fix previews of filtered palettes

Jermolene committed:
Palette tweaks

These filtered palettes are still just experiments with the techniques, and not yet a serious palette

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Update contrastcolour macro to use color.js

Jermolene committed:
Componentise more preview components

Jermolene committed:
Add alert preview

Jermolene committed:
Add notification preview

Jermolene committed:
Fix filtered palette previews

Jermolene committed:
Docs for colour-oklch operator

Jermolene committed:
Start fixing the TwentyTwenties palettes

Jermolene committed:
Add colour-contrast operator

Jermolene committed:
Add contrast checks to TwentyTwenties palette

I think I might be building a programming language for writing palettes...

Jermolene committed:
Display test results at the bottom of the palette switcher

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Fix colour for site title

Jermolene committed:
Make the alert and notification previews be optional

Via some future UI

Jermolene committed:
Refactor for clearer variable names

Jermolene committed:
Refactor colour-oklch operator to colour-set-oklch

Jermolene committed:
Add colour-get-oklch operator

Jermolene committed:
Add colour-interpolate operator

Jermolene committed:
Include contrast value in contrast errors

Jermolene committed:
Barebones docs for colour spaces

Jermolene committed:
Improvements to TwentyTwenties palettes

Still a work in progress, but getting more coherent

kookma commented:
Hi Jeremy,
 Is there updated version of Colour handling to test?

Thank you

pmario commented on editions/tw5.com/tiddlers/filters/colour-get-oklch.tid:
````diff
@@ -0,0 +1,33 @@
+For example, this filter expression will get the lightness of the colour current page background colour:
+
+```
+[function[colour],[page-background]colour-get-oklch:l[]]
````
I did test this: `[function[colour],[page-background]colour-get-oklch:h[]]` -> RSOD See suffix **:h**

<details><summary>Click for screenshot</summary>

![image](https://github.com/user-attachments/assets/c077b1a7-4033-4d8f-a72e-9ab92095b3b5)

</details>

Jermolene replied:
Hi @pmario I'm not able to reproduce this with the latest build. Could you kindly check?

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
TwentyTwenties palette: use interpolation

Jermolene commented:
Hi @kookma 

> Is there updated version of Colour handling to test?

I've been making a steady trickle of changes in the background – this is my current "low priority but fun" project :)

You can see the latest changes here:

https://deploy-preview-8702--tiddlywiki-previews.netlify.app

pmario commented:
@Jermolene -- To reproduce RSOD

- Open: https://deploy-preview-8702--tiddlywiki-previews.netlify.app/
- copy this filter into **AdvancedSearch: Filter tab**: `[function[colour],[page-background]colour-get-oklch:h[]]`

Jermolene committed:
Fix palette switcher

Jermolene committed:
Use interpolation to derive colours

Jermolene committed:
Add a primitive custom palette editor

Jermolene committed:
Add best contrast operator

Jermolene committed:
Fix editing colours that are not in 6 digit hex format

See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/color#value

Jermolene committed:
Download button should use palette colours

Jermolene committed:
Clarify method name

Jermolene committed:
Fix more TwentyTwenties entries

Jermolene committed:
Fix some old-style palette references

Jermolene committed:
Add a palette that automatically switches between dark and light

Will requires #7999 to work fully; for the moment when you switch between dark and light you will see the preview change, and then next time you select the palette it will be in the correct mode

springerspandrel commented:
I haven't followed discussion here before, but came because of Mohammad's recent link ([color picker wish list](https://talk.tiddlywiki.org/t/color-colour-picker-wish-list/11404/6?u=springer))...

In this recent preview (https://deploy-preview-8702--tiddlywiki-previews.netlify.app/) editing the active palette doesn't affect color display in the overall wiki at all (though it does dynamically change the preview thumbnail, which is fun to see). 

If live editing of palettes has been prevented, (for the sake of sandboxing or whatever), or just hasn't been re-constructed in the new build, feel free to ignore my worry!

Jermolene committed:
Refactor actions for recompiling current palette

Jermolene committed:
Palette manager should recompile palette on edits

Jermolene committed:
Refactor TwentyTwenties editor

Jermolene committed:
Edit text widget shouldn't fully refresh when default attribute changes

Jermolene commented:
Thanks @springerspandrel. This is fixed in 69363bf7ef1caa30ae3505b72b66976371fbdc4b

The background is that palettes are now compiled from the raw wikified/filtered text into plain CSS colour values. `$:/palette` continues to point to the current selected palette, and the palette itself can not only use complex wikitext but now modern palettes can use filters to define palette entries. The new twist is that the palette is compiled into `$:/temp/palette-colours` which just contains plain CSS colour values that can be directly transcluded.

Until 69363bf7ef1caa30ae3505b72b66976371fbdc4b  I hadn't yet updated the main control panel palette editor to make it trigger a recompilation. Now that this is fixed the palette editor should work as you expect.

The hope is that now the palette editor would only be needed for people who are creating their own palettes from scratch. Newer palettes can instead be customised using the custom palette editor that shows up for new palettes that elect to use it:

<img width="540" alt="image" src="https://github.com/user-attachments/assets/bcc2642e-d60d-432b-a790-5ad4cdc5e9aa" />

The new TwentyTwenties palette is inspired by the [Flexoki](https://stephango.com/flexoki) palette, except that one can change the paper and ink colours directly, and have the other colours and tones of the palette automatically switch to adjust harmoniously. Try changing the hue of the paper colour:
<img width="1289" alt="image" src="https://github.com/user-attachments/assets/cac85715-255e-4501-b813-2f81b947d3a0" />

Jermolene committed:
Fix palette switcher used in the sidebar

by adding a new "thumbnails" parameter to the template

springerspandrel commented:
Very promising developments. The "ink" and "paper" metaphors are intuitive, and interpolation effects seem to be working well! 

A couple small notes, playing with the TwentyTwenties variations (and as always feel free to ignore if these are already obvious, and/or it's not the right moment in your workflow to address them)

(1) TABS: 

Sidebar tabs are not yet "inheriting" the `tab-background` and `tab-background-selected` palette values in TwentyTwenties (though it seems that's the intention of having these default to `[function[colour],[tab-background]]` (etc.)

(2) PREVIEW THUMBNAILS

(a) `tab-background` and `tab-background-selected` colors don't get any preview even though sidebar tabs are very much part of "out-of-box" GUI... Having a mock set of sidebar tab-rectangles (one active, one or two inactive) might be helpful, at the risk of more visual noise... (Meanwhile, we do currently get a preview of search window input-box — despite duplicating tiddler background for all bundled palettes except Solarized Dark.)  

(b) Selecting new `primary` color doesn't affect palette's preview-thumbnail at all. Selecting new `primary` value does immediately affect the *actual* sidebar-tab list links, and site title (maybe because it's effectively a link, while thumbnail doesn't treat it as a link)... (ONE solution: Include minimalist mock-preview of sidebar tab list content — so the tiddler-link color appears in preview, whether or not a palette has distinct value for site-title-foreground ... ALTERNATE solution: have some primary link-color "accents" within the top mock tiddler such as HelloThere — noting that the real HelloThere conveniently *does* feature several visible links within a short tiddler body!). 

(c) Selecting new `secondary` and/or `tertiary` color choices doesn't affect palette preview, and user *also* lacks obvious places — in larger interface — to recognize dynamic effects of that selection. If the second "mock" tiddler in preview-thumbnail were to include a small block of code, that would nicely gives a hint of `tertiary`. (Including a natural preview of the secondary color seems more difficult, since having an alert within preview thumbnail is awkward.)

Thanks for all this nuanced work!

Jermolene committed:
Paldette and style tweaks

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge background actions and media tracker from #8555

The changes in #8555 are needed in order to be able to offer the desired user experience for dark mode changes.

Jermolene committed:
Make sure rootwidget is available before background actions start

Jermolene committed:
Autocompile palettes when they change

Jermolene committed:
Add a basic dark palette

Obviously some things are hardcoded at the moment

Jermolene committed:
Compile palette at startup

Jermolene committed:
Detect darkmode at startup

Jermolene committed:
Allow text editor type=color for colours not in hex RGB format

See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/color#value

Jermolene committed:
Palette editors no longer need to manually recompile the palette

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Start adding tests for palette operations

Jermolene committed:
Move modern palettes to correct directory

Jermolene committed:
Introduce an improved but temporary cache invalidation method for palettes

Jermolene committed:
Replace accumulate-palette-entries with new changecount filter

Much more direct and efficient

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Allow background actions to be scoped by platform

Jermolene committed:
Remove extraneous logging

Jermolene committed:
Minor cleanups

Jermolene committed:
Proper recursion detection for palettes

Jermolene committed:
Remove infinite loop that was added for testing

Jermolene committed:
Restore accidentally deleted field

Rhys-T commented:
This looks like it will help with a lot of the things I was running into while trying to port the [Catppuccin](https://github.com/catppuccin/catppuccin) palettes to TW. (I actually discovered #8902 while trying to make my own - rather hacky - palette-compiling system on top of the existing palette support.)

> I'm working an "import" field for basing one palette on another. So in the new world, rather than cloning an existing palette, one would subclass it.

Are you planning to support chained `palette-import`s? The use case I'm picturing is something like:

- Palette dev is implementing a set of related palettes with the same 'structure', like the Catppuccin flavors, or the [base16](https://github.com/chriskempson/base16) palettes.
- Dev creates a 'base' dictionary that maps the TW color names to e.g. the Catppuccin ones, but _doesn't_ define the actual color values (and is not tagged `$:/tags/Palette`, because it can't actually be used directly).
- Dev then creates palettes for each flavor that define the Catppuccin color names as actual RGB values, then import the TW color definitions from the base dictionary.
- User tries to create their own tweaked variant of the Mocha flavor, inheriting from the dev's version of it. Currently, this won't import the TW definitions from its 'grandparent', meaning all the colors will go back to the default palette.

It's probably a bit of an edge case, so if recursively scanning `palette-import` like that while avoiding infinite loops is going to be too much of a pain, that's fine. I just wanted to check.

Jermolene committed:
Include tiddler borders in preview

Several of the core palettes have distinctive borders

Jermolene committed:
Introduce dynamic colour scheme mechanism

Also introduces palette inheritance

This finally allows us to have a palette that automatically switches between dark and light variants. The mechanism is more flexible that that, and allows for multiple colour schemes (night, morning, day, evening, for example) with automatic switching between them.

pmario commented on core/wiki/macros/CSS.tid:
```diff
@@ -1,43 +1,60 @@
-\procedure actions-compile-palette-wikified(tempPalette,outputTitle)
+\procedure actions-compile-palette-wikified(tempPalette,outputTitle,scheme)
 <!-- Define our own deeply backwards compatible local versions of the colour macro -->
 \define colour(name)
```
There are 2 `\define` inside the procedure. Is this intentional?

Jermolene replied:
Yes. They are nested definitions. A macro is being used here for backwards compatibility.

Jermolene commented:
> Are you planning to support chained `palette-import`s? The use case I'm picturing is something like:

Yes, I've just implemented that feature in 431149d20c8edfbe580ad88025006105380b3980. I think the approach you've outlined would work.

I'll be updating the OP and the docs as soon as I can.

Jermolene committed:
Typo

Jermolene committed:
Remove logging

Jermolene committed:
Rethink palette manager

The current content of $:/PaletteManager is moved into $:/PaletteEditor, and $:/PaletteManager repurposed as the control panel palette switcher

Jermolene committed:
Compile colour scheme handling

Jermolene committed:
Easier for development if the prerelease uses the tw5.com palette

Jermolene committed:
Set colour scheme property of root element

Jermolene committed:
Clarify comment

Jermolene committed:
Refactor palette switcher a little

Jermolene committed:
Palette should autocompile when the palette tiddler itself changes

Jermolene committed:
The lingo macro shouldn't mess with currentTiddler

Jermolene committed:
Give the AutoToggle palette a custom editor

Jermolene committed:
Rename custom palette editor to custom palette settings

To avoid confusion with existing usage of "palette editor"

Jermolene committed:
Only allow palettes to import palettes of the same type

Jermolene committed:
Refactor some variable names

Jermolene committed:
Remove unneeded scheme processing

Now that we're dealing with the schemes during the import process it is no longer necessary for the compilation process to worry about it.

Jermolene committed:
Get rid of palette-types...

Instead, we'll special case transforming `<<colour X>>` into `[function[colour],[x]]`

Makes everything much easier and avoids all the kerfuffle of not being able to mix palettes

Jermolene committed:
Add VanillaCherry palette

Jermolene committed:
Introduce background palettes for plugins

Jermolene committed:
The tags macros do not actually need access to the palette

Jermolene committed:
Take advantage of compiled palettes

More robust because previously we couldn't cope with indirect palette entries in these situations

Jermolene committed:
Missed off 961b26a984fd93f94b8e2686771bc2e59f03c745

Jermolene committed:
Refactor the palette preview macros into a generic widget

Takes the opportunity to simplify things now that we don't have to worry about wikified palettes

Jermolene committed:
Add tabs to the preview

Jermolene committed:
AutoToggle should default to light mode so that static exports are light mode

Jermolene committed:
Move generic tests into a background palette

These are the generic tests that should be applied to every palette

Jermolene committed:
Typo

Jermolene committed:
Add tabs and recent list to the sidebar

Jermolene committed:
Fill in some missing palette entries

Jermolene committed:
Fix Codemirror colour palette fallbacks

Jermolene committed:
Typo

Jermolene committed:
Add hue adjuster to colour-interpolate operator

Jermolene committed:
Add apply filter run prefix

The map filter run prefix is often used as a way to move a computed value in the input list into a variable so that it can be used as a parameter of a filter operator. The apply filter run prefix extends this idea to make the input list available as variables $1, $2 etc. Unlike the map prefix, the apply filter run is only evaluated once.

Jermolene committed:
Refactor the interpolate operator so it can be used with the range operator

Jermolene committed:
Move sample background item out of the core

It shouldn't be enabled by default, either, but we'll come back to that

Jermolene committed:
Don't need consent banner anymore

Was just for testing

Jermolene committed:
Move from v5.3.7 -> v5.4.0

Jermolene committed:
Docs for apply filter run prefix

Jermolene committed:
Add temporary guide tiddler at the top

Jermolene commented:
> Very promising developments. The "ink" and "paper" metaphors are intuitive, and interpolation effects seem to be working well!

Thanks @springerspandrel 

> Sidebar tabs are not yet "inheriting" the `tab-background` and `tab-background-selected` palette values in TwentyTwenties (though it seems that's the intention of having these default to `[function[colour],[tab-background]]` (etc.)

That should be fixed now.

> (a) `tab-background` and `tab-background-selected` colors don't get any preview even though sidebar tabs are very much part of "out-of-box" GUI... Having a mock set of sidebar tab-rectangles (one active, one or two inactive) might be helpful, at the risk of more visual noise... (Meanwhile, we do currently get a preview of search window input-box — despite duplicating tiddler background for all bundled palettes except Solarized Dark.)

The previews now show sidebar tabs in the appropriate colours.

> (b) Selecting new `primary` color doesn't affect palette's preview-thumbnail at all.

That should be fixed now.

> (ONE solution: Include minimalist mock-preview of sidebar tab list content — so the tiddler-link color appears in preview, whether or not a palette has distinct value for site-title-foreground ... ALTERNATE solution: have some primary link-color "accents" within the top mock tiddler such as HelloThere — noting that the real HelloThere conveniently _does_ feature several visible links within a short tiddler body!).

Both suggestions are now implemented.


> (c) Selecting new `secondary` and/or `tertiary` color choices doesn't affect palette preview

That should be fixed now.

>  and user _also_ lacks obvious places — in larger interface — to recognize dynamic effects of that selection. If the second "mock" tiddler in preview-thumbnail were to include a small block of code, that would nicely gives a hint of `tertiary`.

A block of code is an excellent idea.

> (Including a natural preview of the secondary color seems more difficult, since having an alert within preview thumbnail is awkward.)

I have included an alert in the latest update, but the challenge is that it obscures the title of the first tiddler. Perhaps previews need multiple pages to show things like alerts, notifications and modals.

pmario commented on core/ui/PaletteEditor.tid:
```diff
@@ -0,0 +1,94 @@
+title: $:/PaletteEditor
```
Indentation please.

Jermolene replied:
Just to be clear, this file hasn't been changed (yet), it is just moved from $:/PaletteManager

pmario replied:
ah ok

Jermolene replied:
I meant to add that this file needs substantial revision so I will add indentation when I do that.

Jermolene committed:
Background action demos should require explicit enabling

Jermolene committed:
Introduce tf.colour function to make palette entries more concise

pmario commented on core/wiki/macros/CSS.tid:
```diff
@@ -3,8 +3,9 @@ tags: $:/tags/Macro
 
```
tags: $:/tags/Global is missing

Jermolene replied:
Hi @pmario all the core macros use $:/tags/Macro rather than $:/tags/Global. I think we kept it that way to be super cautious about backwards compatibility. We can discuss changing them, but they should all be changed at the same time.

Leilei332 replied:
IMO we should use `$:/tags/Global` for newly introduced core macros,widgets and functions, while `$:/tags/Macro` for old core macros.

Jermolene replied:
> IMO we should use `$:/tags/Global` for newly introduced core macros,widgets and functions, while `$:/tags/Macro` for old core macros.

That's reasonable. `$:/core/macros/CSS` is of course old, but I may do further refactoring of this PR.

pmario commented:
@Jermolene .. Am I right, that palettes are becoming a "description" how they should be "consolidated" and that they can be chained.

Is this PR ready to be tested or should I wait?

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene commented:
> @Jermolene .. Am I right, that palettes are becoming a "description" how they should be "consolidated" and that they can be chained.

I wouldn't think of it that way.

The consolidation process is just part of the implementation of the palette import process. It's not strictly necessary but it's helpful for debugging to be able to see the final values that were used to compose the palette.

> Is this PR ready to be tested or should I wait?

Do you mean should you click through to the demo? It's probably hard to assess this PR without trying out the demo. It is still a draft, and there's more to do.

Jermolene commented on core/modules/background-actions.js:
```diff
@@ -0,0 +1,129 @@
+
+BackgroundActionDispatcher.prototype.trackFilter = function(title) {
+	var tiddler = this.wiki.getTiddler(title),
+		id = this.nextTrackedFilterId++,
```
Thanks @pmario. These ids are intended to be opaque to client code; there is no semantics in incrementing or decrementing them, they're just used as an identifier.

pmario replied:
ok

Jermolene commented on core/modules/background-actions.js:
```diff
@@ -0,0 +1,129 @@
+	this.wiki = options.wiki;
+	this.title = options.title;
+	this.trackFilter = options.trackFilter;
+	this.actions = options.actions
```
Thanks @pmario

Jermolene commented on core/wiki/macros/tag.tid:
```diff
@@ -28,15 +28,15 @@ color:$(foregroundColor)$;
 \end
 
-\define tag-pill-body(tag,icon,colour,palette,element-tag,element-attributes,actions)
+\define tag-pill-body(tag,icon,colour,element-tag,element-attributes,actions)
```
Thanks @pmario

Jermolene committed:
Fixes suggested by @pmario

Rhys-T commented on core/wiki/paletteswitcher.tid:
```diff
@@ -1,19 +1,30 @@
+
+<$linkcatcher actions=<<actions>>>
+	<div class={{{ tc-chooser [<thumbnails>match[yes]then[tc-chooser-cards]] +[join[ ]] }}}>
+		<$list filter="[all[shadows+tiddlers]tag[$:/tags/Palette]sort[name]]">
```
One minor thing I noticed is that the palette switcher shows any draft versions of palettes, in addition to the actual saved palettes. (This is actually present in the current version too, but there it could plausibly be almost useful if you were trying to edit a palette with the regular tiddler editor and see the changes live. That doesn't work with the new compilation system, though.)

Jermolene replied:
Thanks @Rhys-T I fixed this in master in 3051e8dddc0b4fecee1f9b7fe5895199690718e5

pmario commented:
I did the following tests using Windows 11 pro, FireFox 135

- Selecting [AutoToggle](https://deploy-preview-8702--tiddlywiki-previews.netlify.app/#%24%3A%2Fpalettes%2FAutoToggle) with default settings
 - Change Windows dark / light -> Works
 - Overwrite Setting in FireFox -> Works
 - Change dark / light in FireFox -> Works

------

- Import my [palette-switcher plugin](https://wikilabs.github.io/editions/palette-switcher/) -> Switching works, but it messes up the PR $:/PaletteManager
<details> <summary>See screenshot</summary>

 ![image](https://github.com/user-attachments/assets/f5f651a6-df06-4395-9d07-fab5fd64b356)
</details>

- Import my [palette-watch plugin](https://wikilabs.github.io/editions/palette-switcher/#%24%3A%2Fplugins%2Fwikilabs%2Fpalette-watch), which needs a **save and reload** -> Works
- AutoToggle -> Works
- My Plugins -> Work
- FireFox setting uses my plugin -> Works
- Windows colour settings use my plugin -> Works

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene commented:
Thanks for trying things out @pmario.

> * Import my [palette-switcher plugin](https://wikilabs.github.io/editions/palette-switcher/) -> Switching works, but it messes up the PR $:/PaletteManager

Weird. The preview is showing thumbnails, but `$:/snippets/paletteswitcher` is written to default to omitting the thumbnails, and requires an explicit parameter to do so.

> See screenshot

The fact that the styling in `$:/PaletteManager` has been changed by the palette-switcher plugin suggests that that plugin is redefining core styles, which would usually be considered bad practice.

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Fix from mismerging from master

Jermolene committed:
TwentyTwenties: copy over missing palette entries

These are mostly RGB entries that were previously missing, filled in with values from Vanilla.

The goal is still not to have any direct RGB colours in the palette, just computed colours derived from the base colours

Jermolene committed:
Preview: Fix sidebar link colour

Jermolene committed:
Include palette entry name in palette entries

Makes debugging easier, and works in CSS and as a style.prop assignment

Jermolene committed:
Add colour-set-alpha operator

Jermolene committed:
Fix colour-interpolate with missing colours

Jermolene committed:
TwentyTwenties improvements

Jermolene committed:
Fix tests

Jermolene commented on core/palettes/TwentyTwenties.tid:
```diff
@@ -46,26 +48,26 @@ base-magenta: #CE5D97
-button-border:
+button-background: 
+button-border: 
+button-foreground: 
```
I intend to fill in those entries later.

Jermolene commented on core/palettes/TwentyTwenties.tid:
```diff
@@ -76,7 +78,7 @@ dropdown-background: [tf.colour[background]]
 dropdown-tab-background-selected: [tf.colour[background]]
 dropdown-tab-background: [tf.interpolate-colours[base-paper],[base-ink],[0.9]]
-dropzone-background: rgba(0,200,0,0.7)
+dropzone-background: [tf.colour[base-secondary]colour-set-alpha[0.7]]
```
Hi @pmario to darken a colour one would use `colour-set-oklch:l[0.5]` to set the lightness.

We don't yet have operators to extract the components of a colour in a particular colour space, but once we have those then it would be possible to make macros for setting the alpha or brightness in relative steps.

Jermolene replied:
> Question Does "alpha" set transparency or does it calculate a lighter or darker colour?

Apologies, to be clear it sets transparency

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Merge branch 'master' into colour-improvements

Leilei332 commented on core/modules/background-actions.js:
```diff
@@ -0,0 +1,129 @@
+Class to dispatch actions when filters change
+
+\*/
+(function(){
+
+/*jslint node: true, browser: true */
+/*global $tw: false */
+"use strict";
```
Function wrappers should be removed now.

Jermolene replied:
Thanks @Leilei332 I've removed them in a4c84d727ca45799108763cced1e9abc0f24a809 – I think they still need to be removed from a number of my other open PRs.

Leilei332 commented on core/modules/background-actions.js:
```diff
@@ -0,0 +1,129 @@
+	return id;
+};
+
+BackgroundActionDispatcher.prototype.untrackFilter = function(enterValue) {
```
Since ES2017 is the baseline now, I think we are ready to migrate to  `class` (since refactoring to class is easy in VSCode). We should also use function argument default value and use `const` in some occasions.

Jermolene replied:
Good point, which again applies to my other PRs.

Jermolene commented:
We need to discuss the considerable increase in size of the core incurred by this PR. Nearly 300KB is a considerable impact; I think it's 11% of the current size of empty.html.

> 📊 Build Size Comparison: empty.html
> 
> Branch	Size
> Base (master)	2529.5 KB
> PR	2822.2 KB
> Diff: ⬆️ Increase: +292.6 KB

My original thinking was that we could conceal the size impact by adopting compression for the core, but it would be reasonable to expect that the percentage increase in size would be roughly similar.

I wondered if it might be possible to reduce the size of the library through tree shaking. I don't have experience of the tooling but perhaps others might be able to comment. However, we are using some substantial subsystems from the library (such as parsing CSS colours, which I would imagine might be a hefty chunk of code), and so the impact might not be enormous.

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Remove function wrappers

Jermolene committed:
Adopt class syntax

Jermolene committed:
Adopt more ES2017 features

Jermolene committed:
Change default colour space for interpolation

Jermolene committed:
Add new palettes from https://yatagarasu.tiddlyhost.com

Jermolene committed:
Move palettes into category folders

Jermolene committed:
Add category field to palettes

Leilei332 commented:
Some tags' foreground colors looks weird to me.

![image](https://github.com/user-attachments/assets/9f556d29-4867-4cb9-ab66-48e65afb29c2)

Leilei332 commented:
> We need to discuss the considerable increase in size of the core incurred by this PR. Nearly 300KB is a considerable impact; I think it's 11% of the current size of empty.html.
> 
> 
> 
> > 📊 Build Size Comparison: empty.html
> 
> > 
> 
> > Branch	Size
> 
> > Base (master)	2529.5 KB
> 
> > PR	2822.2 KB
> 
> > Diff: ⬆️ Increase: +292.6 KB
> 
> 
> 
> My original thinking was that we could conceal the size impact by adopting compression for the core, but it would be reasonable to expect that the percentage increase in size would be roughly similar.
> 
> 
> 
> I wondered if it might be possible to reduce the size of the library through tree shaking. I don't have experience of the tooling but perhaps others might be able to comment. However, we are using some substantial subsystems from the library (such as parsing CSS colours, which I would imagine might be a hefty chunk of code), and so the impact might not be enormous.
> 
> 
> 
> 
> 
> 

IMO we should use minified version of colorjs here, the original one includes many JSDoc comments, which caused the library very large. And since v5.4.0 supports ES2017, we can use the latest version. I think it should be replaced by the latest minified version: https://colorjs.io/dist/color.global.min.js

Jermolene committed:
Switch to minified version of color.js

See this discussion https://github.com/TiddlyWiki/TiddlyWiki5/pull/8702#issuecomment-3158465038

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene commented:
Thanks @Leilei332 I've updated to the minified version in 974904588cd3a345d3fbb8f0ec454e1dc8049be2 but it appears to have had much less impact than I'd hoped. If I'm reading things correctly, the move saves 9KB.

Jermolene committed:
Use TwentyTwenties palette by default

Jermolene committed:
Prepare to split palettes out into plugins

Jermolene committed:
Move palettes into plugins

Jermolene committed:
Merge branch 'master' into colour-improvements

BurningTreeC commented:
Hello @Jermolene - [here](https://github.com/TiddlyWiki/TiddlyWiki5/pull/9216#issuecomment-3154299257) you mentioned you will merge the mediaquery tracking as soon as possible. Is it ready? I would need it for #9507 
☺️

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Avoid using camel case for new docs tiddlers

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Upgrade color.js

Jermolene committed:
Typo

Jermolene committed:
Switch from apply filter run to let filter run

Jermolene committed:
Don't lint color.js

Jermolene committed:
Enable change note

Jermolene committed:
Update the 2025 palette plugin to 2026

Jermolene committed:
Remove docs for apply filter run prefix

Jermolene committed:
Update new palettes to 2026 category

Jermolene committed:
Properly credit the new palettes

Jermolene committed:
Fix lint errors

Jermolene committed:
More linting errors

Jermolene committed:
More linting dancing

Jermolene committed:
Remove debugging code

It was breaking colour assignment as a property, rather than in a stylesheet

Jermolene committed:
Fix remaining references to palettes-2025 plugin

Jermolene committed:
Revert linter fixes for unused function parameters

Jermolene committed:
Fix year category for new palettes

Jermolene committed:
Fix use of creaky old technique for setting alpha/opacity from a different base colour

Jermolene committed:
Tweak credits for 2026 palettes

Jermolene committed:
Merge branch 'master' into colour-improvements

pmario commented on core/modules/filter-tracker.js:
```diff
@@ -0,0 +1,106 @@
+	fnProcess: function to call each time the tracker is processed, after any enter, leave or change functions are called. Called as (changes)
+	*/
+	track(options = {}) {
+		const {
```
I found this one too. Will send a PR soon.

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Fix tests

Jermolene committed:
Allow default filter run prefix to be specified for filter evaluation

Jermolene committed:
Extend subfilter operator with suffix for specifying the default filter run prefix

Jermolene committed:
temp

Jermolene committed:
Merge subfilter fix from #9595

Jermolene committed:
Use the subfilter suffix to suppress deduping palette entries

Jermolene committed:
Merge branch 'default-prefix-for-subfilter' into colour-improvements

Jermolene committed:
Add a P3 palette for testing

Jermolene committed:
Add a view template body for palettes

Jermolene committed:
Revert previous merge from #9595

This reverts commit eb8f69dd038562adbc941b95a8125ff880010e6d.

Jermolene committed:
Merge branch 'default-prefix-for-subfilter' into colour-improvements

Jermolene committed:
Fix parsing bug

Jermolene committed:
Add test for invalid pragma

Jermolene committed:
Improve English

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Tweak lingo

Jermolene committed:
Remove appply filter run prefix

Jermolene committed:
Make Palette be the default Appearance tab in control panel

Jermolene committed:
Impact note about wikified palettes

Leilei332 commented:
I'm afraid that this PR may be too little huge to be completed before 6th Feb. I think we may split these features to be included in v5.4.0:

* media query tracker
* `apply` filter run prefix

pmario commented:
**Just as a reference**: https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/color_value/contrast-color 

@Jermolene ... It's a 2025 CSS feature supported by most browsers. We may be able to use it and fall back to the new library functions.

Jermolene committed:
Merge branch 'master' into colour-improvements

Jermolene committed:
Linting

Jermolene committed:
Typo

Jermolene committed:
Default to the hire jeremy banner in the sidebar

Jermolene committed:
Only initialise performance framework once

Jermolene committed:
Guard against  an unknown default filter run prefix

Jermolene committed:
Improve introduction

Jermolene committed:
Remove typo

Jermolene committed:
Update release note

Jermolene committed:
Fix the old, deprecated parseCSSColorObject function to return 0-255

Jermolene committed:
Update release notes

Jermolene commented:
> ## ❌ Change Note Status
> Change note validation failed. Please fix the following issues:

Hi @saqimtiaz is this correct? Our current docs just specify that the suffix should be unique: https://tiddlywiki.com/#Release%20Notes%20and%20Changes%20Internals

Jermolene committed:
Fix palette typo

Jermolene committed:
Another palette typo

Jermolene committed:
Docs

Jermolene committed:
Fix incorrect signature for calling wiki.filterTiddlers method

Jermolene committed:
Fix compilation of scheme-suffixed palette entries

Suffixed entries are now copied after unsuffixed entries so that they override
them as documented, the names and values are sorted consistently, and a
malformed <%endif%> is corrected

Jermolene committed:
Fix palette preview handling of scheme-suffixed entries and circular imports

The scheme-suffixed lookup key was constructed backwards, and imported
palettes are now tracked in an exclusions list to guard against loops

Jermolene committed:
Report palette entries with recursive definitions

Recursive definitions are blanked and reported via the readability test mechanism

Jermolene committed:
Fix editor colour scheme lookup and refreshing of the default attribute

Jermolene committed:
Guard colour operators against achromatic hues and unparseable target colours

Jermolene committed:
Fix convertCSSColorToRGBString to always return a six digit hex value

Jermolene committed:
Ship the Vanilla palette in the core for backwards compatibility

Also fixes typos in the Vanilla Cherry and TwentyTwenties palettes

Jermolene committed:
Fix reference to current palette

Hardcoding was a bug

Jermolene committed:
Pass the palette to the tag-body

Jermolene committed:
Recompile palette on changes to all tagged palette tiddlers

Jermolene committed:
Fix test for recursive palettes

Jermolene committed:
Fix default colour space for colour-interpolate filter

saqimtiaz commented:
> > ## ❌ Change Note Status
> > Change note validation failed. Please fix the following issues:
> 
> Hi @saqimtiaz is this correct? Our current docs just specify that the suffix should be unique: https://tiddlywiki.com/#Release%20Notes%20and%20Changes%20Internals

@linonetwo could you take a look at this please?

Jermolene committed:
Docs updates

Jermolene committed:
Add colour-composite operator and use it to fix contrast tests for translucent colours

Jermolene committed:
Fix tour plugin palette typo

Jermolene committed:
Refactor 2026 palettes to use inheritance

Jermolene committed:
Remove the unimplemented DeltaPhi suffix from colour contrast callers

Jermolene committed:
Clean up palette and filter run prefix code

Removes the unused actions-switch-colour-palette procedure, fixes the filter
and subfilter operators to fall back to the default filter run prefix when
given an empty suffix, and notes that the DefaultColourMappings fallback is
retained for third party compatibility

Jermolene committed:
Impact notes for the colour macro alpha suffix idiom and the lingo macro

Jermolene committed:
Restrict filter pragmas to the start of a filter

Pragmas appearing after a filter run now return an error. This removes the
ambiguity where the subfilter and filter operators saw the final pragma value
regardless of their position. Adds generic docs for filter pragmas and an
impact note for titles beginning with double colons

Jermolene committed:
Lighten TwentyTwenties page background to pass its own contrast test

Jermolene committed:
Document the palette plugins

Adds an impact note for the palettes moving to the Legacy Palettes plugin,
readmes for both palette plugins, and official plugin doc tiddlers on tw5.com

Jermolene committed:
Fall back to the Vanilla palette when the current palette is missing

Jermolene committed:
Fix palette fallback wrongly treating shadow palettes as missing

The is[missing] operator returns true for shadow tiddlers, so the fallback
introduced for missing palettes made every plugin palette compile as Vanilla

Jermolene committed:
Standardise on customisations terminology for palette customisation panels

Renames the settings field to customisations, the panel tiddlers from
Settings to Customisations, and updates the docs to match

Jermolene committed:
Store palette customisations in a layer above the palette itself

Customisations no longer modify the palette tiddler. They are stored in a
tiddler titled $:/config/PaletteCustomisations/<palette-title> that imports
the palette and is compiled in its place, so plugin palettes continue to
receive updates. Adds a reset button to the palette chooser, and a warning
banner and per-entry markers to the palette editor. Also excludes the type
field from palette field inheritance, which previously broke compilation of
JSON data tiddler palettes

Jermolene committed:
Fix customisations panel selection when the palette redirects its imports

The panel is now chosen from the current palette's own customisations field
before falling back to the consolidated palette, so AutoToggle keeps its
panel after a palette has been chosen in it. Also mirrors the colour scheme
inheritance hop in the palette preview, and removes two leftover DeltaPhi
suffixes

Jermolene committed:
Docs corrections from review

Updates the palette customisation docs for the layer model, fixes a mangled
sentence in the change note, and tidies wording

Jermolene committed:
Add tests for changecount, scheme imports, derived palette entries and readability tests

Jermolene committed:
Add missing from-version markers to colour docs

Jermolene committed:
Add tests for colour utilities, sendmessage duplicate values and interpolation suffixes

Jermolene committed:
Increase browser test timeout for slower CI runners

Jermolene committed:
Merge branch 'master' into colour-improvements

pmario commented:
@Jermolene ... Most of the new palettes have a problem with dropdown hover readability. 

<img width="1552" height="328" alt="image" src="https://github.com/user-attachments/assets/576e896d-1e42-4e08-9bc4-c6940cdf5637" />
